File: dotnet/Core/Map/Terrain.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using LumberjackRL.Core.Utilities;

namespace LumberjackRL.Core.Map
{
    public class Terrain : IStoreObject
    {
        private TerrainCode code;
        private int water;  //how much water is here?
        private Dictionary<TerrainParameter,String> parameters;

        public Terrain()
        {
            code = TerrainCode.DIRT;
            water = 0;
            parameters = new Dictionary<TerrainParameter,String>();
        }

        public void SaveObject(StreamWriter outStream)
        {
            outStream.WriteLine(EnumUtil.EnumName<TerrainCode>(code));
            outStream.WriteLine(getWater().ToString());
            outStream.WriteLine(getParameters().Count.ToString());
            foreach(TerrainParameter tempKey in getParameters().Keys)
            {
                outStream.WriteLine(EnumUtil.EnumName<TerrainParameter>(tempKey));
                outStream.WriteLine(parameters[tempKey]);
            }
        }

        public void LoadObject(StreamReader inStream)
        {
            code = (TerrainCode)Enum.Parse(typeof(TerrainCode), inStream.ReadLine());
            setWater(Int32.Parse(inStream.ReadLine()));
            getParameters().Clear();
            int parameterSize = Int32.Parse(inStream.ReadLine());
            for (int i = 0; i < parameterSize; i++)
            {
                TerrainParameter newKey = (TerrainParameter)Enum.Parse(typeof(TerrainParameter),inStream.ReadLine());
                String newValue = inStream.ReadLine();
                getParameters().Add(newKey, newValue);
            }
        }

        /**
         * @return the code
         */
        public TerrainCode getCode() {
            return code;
        }

        /**
         * @return the parameters
         */
        public Dictionary<TerrainParameter, String> getParameters() {
            return parameters;
        }

        /**
         * @param code the code to set
         */
        public void setCode(TerrainCode code) {
            this.code = code;
        }

        /**
         * @return the water
         */
        public int getWater() {
            return water;
        }

        /**
         * @param water the water to set
         */
        public void setWater(int water) {
            this.water = water;
        }
    }
}

File: java/src/quinoa/region/Terrain.java
package quinoa.region;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.util.HashMap;
import quinoa.Storable;
import quinoa.region.TerrainManager.TerrainCode;
import quinoa.region.TerrainManager.TerrainParameter;

public class Terrain implements Storable
{
    private TerrainCode code;
    private int water;  //how much water is here?
    private HashMap<TerrainParameter,String> parameters;

    public Terrain()
    {
        code = TerrainCode.DIRT;
        water = 0;
        parameters = new HashMap<TerrainParameter,String>();
    }

    public void save(BufferedWriter out) throws Exception
    {
        out.write(getCode().name()); out.newLine();
        out.write(getWater() + ""); out.newLine();
        out.write(getParameters().size() + ""); out.newLine();
        for(Object tempKey : getParameters().keySet().toArray())
        {
            out.write(((TerrainParameter)tempKey).name()); out.newLine();
            out.write((String)getParameters().get((TerrainParameter)tempKey)); out.newLine();
        }
    }

    public void load(BufferedReader in) throws Exception
    {
        setCode(TerrainCode.valueOf(in.readLine()));
        setWater(Integer.parseInt(in.readLine()));
        getParameters().clear();
        int parameterSize = Integer.parseInt(in.readLine());
        for(int i=0; i < parameterSize; i++)
        {
            TerrainParameter newKey = TerrainParameter.valueOf(in.readLine());
            String newValue = in.readLine();
            getParameters().put(newKey, newValue);
        }
    }

    /**
     * @return the code
     */
    public TerrainCode getCode() {
        return code;
    }

    /**
     * @return the parameters
     */
    public HashMap<TerrainParameter, String> getParameters() {
        return parameters;
    }

    /**
     * @param code the code to set
     */
    public void setCode(TerrainCode code) {
        this.code = code;
    }

    /**
     * @return the water
     */
    public int getWater() {
        return water;
    }

    /**
     * @param water the water to set
     */
    public void setWater(int water) {
        this.water = water;
    }
}
